[PGNCCL] Fix P2P data corruption in non-blocking mode

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138527
* __->__ #138860

In non-blocking mode, it seems a single `ncclRecv` or `ncclSend` call can "early return" `ncclSuccess` before the kernel is fully enqueued. This causes the event record below missing the P2P the kernel, leading to data corruption.

Side note: per NCCL, it is legal to call `ncclSend` or `ncclRecv` only if there is only one P2P op. This is true whether we are in blocking or non-blocking mode.

In this fix, we use ncclGroup semantics to ensure that the kernel is enqueued for single-P2P ops. The ncclGroup call itself should introduce minimal overhead.

Added a test `test_non_blocking_p2p`.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[PGNCCL] Fix P2P data corruption in non-blocking mode

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3337,10 +3337,14 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
-      ncclComm->getNcclComm(),
-      ncclComm->getNcclCommFailureReason());
+  // In non-blocking mode, we need to use ncclGroup semantics to ensure that the
+  // kernel is enqueued for single-P2P ops.  Otherwise, the event record below
```
i don't understand this explanation.  
 - why is this specific to p2p ops?  when in non-blocking mode, do all collectives need to be in a 'group' to ensure they are correctly enqueued before the subsequent event record?

kwen2501 replied:
I feel we would need better contract / documentation from NCCL in terms of what `ncclSuccess` means for P2P calls -- i.e. whether kernel is guaranteed to have been enqueued or not. In this instant case, it seems not, so a group call is used to help it.

kwen2501 replied:
Re why specific to P2P ops:
P2P ops involves dynamic connection establishment. This is the part made non-blocking in non-blocking mode.

wconstab replied:
hmm, ok that kinda makes sense. 

however i have some concerns still: this change would effectively turn all P2P ops into grouped-P2P ops of size 1.  Does grouping add overhead to P2P performance that could be avoided in the blocking case?  (Overhead either on CPU side or on GPU side- does a 'group' p2p run a different kernel for example?

kwen2501 replied:
No difference on GPU side.
Re CPU side, I wouldn't say there are not more instructions, but the overhead is minimal.

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3337,10 +3337,14 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
+  // kernel is enqueued for single-P2P ops.  Otherwise, the event record below
+  // may not capture the kernel, leading to data corruption.
+  ncclGroupStart();
+  C10D_NCCL_CHECK_NONBLOCKING(
```
Should this MACRO be applied when non-blocking mode is on? otherwise we should just use           C10D_NCCL_CHECK

kwen2501 replied:
I don't quite understand the question. We are in the `NCCL_HAS_COMM_NONBLOCKING` region, and we are using the `C10D_NCCL_CHECK_NONBLOCKING` macro.

shuqiangzhang replied:
Sorry, I mean why not use C10D_NCCL_CHECK_TIMEOUT to wait for ncclSuccess, same as those used in other  collectives

kwen2501 replied:
Oh, the TIMEOUT check is deferred to the `ncclGroupEnd()` call. I think that's more efficient.

kwen2501 replied:
i.e. we don't have timeout concern for calls within group semantics.

kwen2501 committed:
Update on "[PGNCCL] Fix P2P data corruption in non-blocking mode"


In non-blocking mode, we need to use ncclGroup semantics to ensure that the kernel is enqueued for single-P2P ops.  Otherwise, the event record below may not capture the kernel, leading to data corruption.

Added a test `test_non_blocking_p2p`.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Fix P2P data corruption in non-blocking mode"


In non-blocking mode, we need to use ncclGroup semantics to ensure that the kernel is enqueued for single-P2P ops.  Otherwise, the event record below may not capture the kernel, leading to data corruption.

Added a test `test_non_blocking_p2p`.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

eqy commented:
CC @KaimingOuyang for comment

kwen2501 commented:
@KaimingOuyang confirmed that there is a bug in NCCL re `ncclRecv` or `ncclSend`: if we don't poll on `ncclCommGetAsyncError`, the kernel may not be fully enqueued -- which sounds like `ncclRecv` can directly return `ncclSuccess` instead of `ncclInProgress`.
I personally feel that having a group call here is future-proof, in that even after NCCL fix the bug, we don't have to take the group call out; and the group call overhead is minimal.
Cc: @eqy

kwen2501 commented:
@pytorchbot merge
